File: FrontEnd/src/services/publicationService.ts
import { httpClient } from '../utils/api/http-client';

export interface PublicationAuthor {
  id: string;
  name: string;
  email: string;
}

export interface Publication {
  id: string;
  author: PublicationAuthor;
  title: string;
  description: string;
  content: string;
  tags: string[];
  status: 'DRAFT' | 'PUBLISHED' | 'ARCHIVED';
  viewCount: number;
  likeCount: number;
  commentCount: number;
  featuredImage?: string;
  readingTime?: number;
  createdAt: string;
  updatedAt: string;
  publishedAt?: string;
  isLikedByCurrentUser: boolean;
}

export interface CreatePublicationDto {
  title: string;
  description?: string;
  content: string;
  tags?: string[];
  status?: 'DRAFT' | 'PUBLISHED' | 'ARCHIVED';
  featuredImage?: string;
  readingTime?: number;
}

export interface UpdatePublicationDto {
  title?: string;
  description?: string;
  content?: string;
  tags?: string[];
  status?: 'DRAFT' | 'PUBLISHED' | 'ARCHIVED';
  featuredImage?: string;
  readingTime?: number;
}

export interface PublicationPageResponse<T> {
  content: T[];
  totalElements: number;
  totalPages: number;
  size: number;
  number: number;
}

const BASE_URL = '/api/community/publications';

class PublicationService {
  async createPublication(data: CreatePublicationDto): Promise<Publication> {
    const response = await httpClient.post<{ status: string; message: string; data: Publication }>(
      BASE_URL,
      data
    );
    return response.data;
  }

  async getAllPublications(params?: {
    page?: number;
    size?: number;
    sortBy?: string;
    sortDir?: 'ASC' | 'DESC';
  }): Promise<PublicationPageResponse<Publication>> {
    const queryParams = new URLSearchParams();
    if (params?.page !== undefined) queryParams.append('page', params.page.toString());
    if (params?.size !== undefined) queryParams.append('size', params.size.toString());
    if (params?.sortBy) queryParams.append('sortBy', params.sortBy);
    if (params?.sortDir) queryParams.append('sortDir', params.sortDir);
    
    const url = queryParams.toString() ? `${BASE_URL}?${queryParams}` : BASE_URL;
    const response = await httpClient.get<{ status: string; message: string; data: PublicationPageResponse<Publication> }>(url);
    return response.data;
  }

  async getPublicationById(id: string): Promise<Publication> {
    const response = await httpClient.get<{ status: string; message: string; data: Publication }>(
      `${BASE_URL}/${id}`
    );
    return response.data;
  }

  async updatePublication(id: string, data: UpdatePublicationDto): Promise<Publication> {
    const response = await httpClient.put<{ status: string; message: string; data: Publication }>(
      `${BASE_URL}/${id}`,
      data
    );
    return response.data;
  }

  async deletePublication(id: string): Promise<void> {
    await httpClient.delete(`${BASE_URL}/${id}`);
  }

  async toggleLike(id: string): Promise<Publication> {
    const response = await httpClient.post<{ status: string; message: string; data: Publication }>(
      `${BASE_URL}/${id}/like`
    );
    return response.data;
  }

  async searchPublications(params: {
    query: string;
    page?: number;
    size?: number;
  }): Promise<PublicationPageResponse<Publication>> {
    const queryParams = new URLSearchParams();
    queryParams.append('query', params.query);
    if (params.page !== undefined) queryParams.append('page', params.page.toString());
    if (params.size !== undefined) queryParams.append('size', params.size.toString());
    
    const response = await httpClient.get<{ status: string; message: string; data: PublicationPageResponse<Publication> }>(
      `${BASE_URL}/search?${queryParams}`
    );
    return response.data;
  }

  async getPublicationsByTag(params: {
    tag: string;
    page?: number;
    size?: number;
  }): Promise<PublicationPageResponse<Publication>> {
    const { tag, ...rest } = params;
    const queryParams = new URLSearchParams();
    if (rest.page !== undefined) queryParams.append('page', rest.page.toString());
    if (rest.size !== undefined) queryParams.append('size', rest.size.toString());
    
    const url = queryParams.toString() 
      ? `${BASE_URL}/tag/${encodeURIComponent(tag)}?${queryParams}`
      : `${BASE_URL}/tag/${encodeURIComponent(tag)}`;
    const response = await httpClient.get<{ status: string; message: string; data: PublicationPageResponse<Publication> }>(url);
    return response.data;
  }

  async getTrendingPublications(params?: {
    page?: number;
    size?: number;
  }): Promise<PublicationPageResponse<Publication>> {
    const queryParams = new URLSearchParams();
    if (params?.page !== undefined) queryParams.append('page', params.page.toString());
    if (params?.size !== undefined) queryParams.append('size', params.size.toString());
    
    const url = queryParams.toString() ? `${BASE_URL}/trending?${queryParams}` : `${BASE_URL}/trending`;
    const response = await httpClient.get<{ status: string; message: string; data: PublicationPageResponse<Publication> }>(url);
    return response.data;
  }

  async getPublicationsByAuthor(params: {
    email: string;
    page?: number;
    size?: number;
  }): Promise<PublicationPageResponse<Publication>> {
    const { email, ...rest } = params;
    const queryParams = new URLSearchParams();
    if (rest.page !== undefined) queryParams.append('page', rest.page.toString());
    if (rest.size !== undefined) queryParams.append('size', rest.size.toString());
    
    const url = queryParams.toString()
      ? `${BASE_URL}/author/${encodeURIComponent(email)}?${queryParams}`
      : `${BASE_URL}/author/${encodeURIComponent(email)}`;
    const response = await httpClient.get<{ status: string; message: string; data: PublicationPageResponse<Publication> }>(url);
    return response.data;
  }

  async getMyPublications(params?: {
    status?: 'DRAFT' | 'PUBLISHED' | 'ARCHIVED';
    page?: number;
    size?: number;
  }): Promise<PublicationPageResponse<Publication>> {
    const queryParams = new URLSearchParams();
    if (params?.status) queryParams.append('status', params.status);
    if (params?.page !== undefined) queryParams.append('page', params.page.toString());
    if (params?.size !== undefined) queryParams.append('size', params.size.toString());
    
    const url = queryParams.toString() ? `${BASE_URL}/my?${queryParams}` : `${BASE_URL}/my`;
    const response = await httpClient.get<{ status: string; message: string; data: PublicationPageResponse<Publication> }>(url);
    return response.data;
  }
}

const publicationService = new PublicationService();
export default publicationService;

File: Backend/src/main/java/com/enit/satellite_platform/modules/community/services/PublicationService.java
package com.enit.satellite_platform.modules.community.services;

import com.enit.satellite_platform.modules.community.dto.CreatePublicationDto;
import com.enit.satellite_platform.modules.community.dto.PublicationResponseDto;
import com.enit.satellite_platform.modules.community.dto.UpdatePublicationDto;
import com.enit.satellite_platform.modules.community.entities.Publication;
import com.enit.satellite_platform.modules.community.entities.Publication.PublicationStatus;
import com.enit.satellite_platform.modules.community.repositories.PublicationRepository;
import com.enit.satellite_platform.modules.user_management.management_cvore_service.entities.User;
import com.enit.satellite_platform.modules.user_management.normal_user_service.repositories.UserRepository;
import org.bson.types.ObjectId;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;
import org.springframework.security.access.AccessDeniedException;
import org.springframework.security.core.userdetails.UsernameNotFoundException;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.Date;

/**
 * Service class for managing publications.
 * Provides methods for creating, retrieving, updating, deleting publications,
 * and handling engagement features like likes and views.
 */
@Service
public class PublicationService {

    private static final Logger logger = LoggerFactory.getLogger(PublicationService.class);

    @Autowired
    private PublicationRepository publicationRepository;

    @Autowired
    private UserRepository userRepository;

    /**
     * Creates a new publication.
     */
    @Transactional
    public PublicationResponseDto createPublication(CreatePublicationDto createDto, String email) {
        logger.info("Creating publication with title: {} for user: {}", createDto.getTitle(), email);

        User author = userRepository.findByEmail(email)
                .orElseThrow(() -> new UsernameNotFoundException("User not found with email: " + email));

        Publication publication = Publication.builder()
                .author(author)
                .title(createDto.getTitle())
                .description(createDto.getDescription())
                .content(createDto.getContent())
                .tags(createDto.getTags())
                .status(createDto.getStatus())
                .featuredImage(createDto.getFeaturedImage())
                .readingTime(createDto.getReadingTime())
                .viewCount(0L)
                .likeCount(0L)
                .commentCount(0L)
                .createdAt(new Date())
                .updatedAt(new Date())
                .deleted(false)
                .build();

        if (publication.getStatus() == PublicationStatus.PUBLISHED) {
            publication.setPublishedAt(new Date());
        }

        Publication savedPublication = publicationRepository.save(publication);
        logger.info("Publication created successfully with ID: {}", savedPublication.getId());

        return convertToResponseDto(savedPublication, email);
    }

    /**
     * Gets all published publications.
     */
    public Page<PublicationResponseDto> getAllPublished(Pageable pageable, String userEmail) {
        logger.info("Fetching all published publications");
        Page<Publication> publications = publicationRepository.findAllPublished(pageable);
        return publications.map(pub -> convertToResponseDto(pub, userEmail));
    }

    /**
     * Gets publications by author.
     */
    public Page<PublicationResponseDto> getPublicationsByAuthor(String authorEmail, Pageable pageable, String userEmail) {
        logger.info("Fetching publications for author: {}", authorEmail);
        User author = userRepository.findByEmail(authorEmail)
                .orElseThrow(() -> new UsernameNotFoundException("Author not found: " + authorEmail));
        Page<Publication> publications = publicationRepository.findByAuthorIdAndDeletedFalse(author.getId().toString(), pageable);
        return publications.map(pub -> convertToResponseDto(pub, userEmail));
    }

    /**
     * Gets a publication by ID and increments view count.
     */
    @Transactional
    public PublicationResponseDto getPublicationById(String publicationId, String userEmail) {
        logger.info("Fetching publication: {}", publicationId);
        Publication publication = publicationRepository.findByIdAndDeletedFalse(new ObjectId(publicationId))
                .orElseThrow(() -> new RuntimeException("Publication not found: " + publicationId));
        publication.setViewCount(publication.getViewCount() + 1);
        publicationRepository.save(publication);
        return convertToResponseDto(publication, userEmail);
    }

    /**
     * Updates a publication.
     */
    @Transactional
    public PublicationResponseDto updatePublication(String publicationId, UpdatePublicationDto updateDto, String email) {
        logger.info("Updating publication: {}", publicationId);
        Publication publication = publicationRepository.findByIdAndDeletedFalse(new ObjectId(publicationId))
                .orElseThrow(() -> new RuntimeException("Publication not found: " + publicationId));

        User currentUser = userRepository.findByEmail(email)
                .orElseThrow(() -> new UsernameNotFoundException("User not found: " + email));

        if (!publication.getAuthor().getId().equals(currentUser.getId())) {
            throw new AccessDeniedException("You can only update your own publications");
        }

        if (updateDto.getTitle() != null) publication.setTitle(updateDto.getTitle());
        if (updateDto.getDescription() != null) publication.setDescription(updateDto.getDescription());
        if (updateDto.getContent() != null) publication.setContent(updateDto.getContent());
        if (updateDto.getTags() != null) publication.setTags(updateDto.getTags());
        if (updateDto.getFeaturedImage() != null) publication.setFeaturedImage(updateDto.getFeaturedImage());
        if (updateDto.getReadingTime() != null) publication.setReadingTime(updateDto.getReadingTime());
        if (updateDto.getStatus() != null) {
            PublicationStatus oldStatus = publication.getStatus();
            publication.setStatus(updateDto.getStatus());
            if (oldStatus != PublicationStatus.PUBLISHED && updateDto.getStatus() == PublicationStatus.PUBLISHED) {
                publication.setPublishedAt(new Date());
            }
        }

        publication.setUpdatedAt(new Date());
        Publication updatedPublication = publicationRepository.save(publication);
        return convertToResponseDto(updatedPublication, email);
    }

    /**
     * Deletes a publication (soft delete).
     */
    @Transactional
    public void deletePublication(String publicationId, String email) {
        logger.info("Deleting publication: {}", publicationId);
        Publication publication = publicationRepository.findByIdAndDeletedFalse(new ObjectId(publicationId))
                .orElseThrow(() -> new RuntimeException("Publication not found: " + publicationId));

        User currentUser = userRepository.findByEmail(email)
                .orElseThrow(() -> new UsernameNotFoundException("User not found: " + email));

        if (!publication.getAuthor().getId().equals(currentUser.getId())) {
            throw new AccessDeniedException("You can only delete your own publications");
        }

        publication.setDeleted(true);
        publication.setDeletedAt(new Date());
        publicationRepository.save(publication);
    }

    /**
     * Toggles like on a publication.
     */
    @Transactional
    public PublicationResponseDto toggleLike(String publicationId, String email) {
        logger.info("Toggling like on publication: {}", publicationId);
        Publication publication = publicationRepository.findByIdAndDeletedFalse(new ObjectId(publicationId))
                .orElseThrow(() -> new RuntimeException("Publication not found: " + publicationId));

        User user = userRepository.findByEmail(email)
                .orElseThrow(() -> new UsernameNotFoundException("User not found: " + email));

        String userId = user.getId().toString();
        if (publication.getLikedBy().contains(userId)) {
            publication.getLikedBy().remove(userId);
            publication.setLikeCount(publication.getLikeCount() - 1);
        } else {
            publication.getLikedBy().add(userId);
            publication.setLikeCount(publication.getLikeCount() + 1);
        }

        Publication updated = publicationRepository.save(publication);
        return convertToResponseDto(updated, email);
    }

    /**
     * Searches publications by title.
     */
    public Page<PublicationResponseDto> searchPublications(String query, Pageable pageable, String userEmail) {
        Page<Publication> publications = publicationRepository.searchByTitle(query, pageable);
        return publications.map(pub -> convertToResponseDto(pub, userEmail));
    }

    /**
     * Gets publications by tag.
     */
    public Page<PublicationResponseDto> getPublicationsByTag(String tag, Pageable pageable, String userEmail) {
        Page<Publication> publications = publicationRepository.findByTagAndPublished(tag, pageable);
        return publications.map(pub -> convertToResponseDto(pub, userEmail));
    }

    /**
     * Gets trending publications.
     */
    public Page<PublicationResponseDto> getTrendingPublications(Pageable pageable, String userEmail) {
        Page<Publication> publications = publicationRepository.findTrending(pageable);
        return publications.map(pub -> convertToResponseDto(pub, userEmail));
    }

    /**
     * Gets current user's publications.
     */
    public Page<PublicationResponseDto> getMyPublications(String email, PublicationStatus status, Pageable pageable) {
        User user = userRepository.findByEmail(email)
                .orElseThrow(() -> new UsernameNotFoundException("User not found: " + email));

        Page<Publication> publications;
        if (status != null) {
            publications = publicationRepository.findByAuthorIdAndStatus(user.getId().toString(), status, pageable);
        } else {
            publications = publicationRepository.findByAuthorIdAndDeletedFalse(user.getId().toString(), pageable);
        }
        return publications.map(pub -> convertToResponseDto(pub, email));
    }

    private PublicationResponseDto convertToResponseDto(Publication publication, String userEmail) {
        boolean isLiked = false;
        if (userEmail != null) {
            User user = userRepository.findByEmail(userEmail).orElse(null);
            if (user != null) {
                isLiked = publication.getLikedBy().contains(user.getId().toString());
            }
        }

        return PublicationResponseDto.builder()
                .id(publication.getId().toString())
                .author(PublicationResponseDto.AuthorDto.builder()
                        .id(publication.getAuthor().getId().toString())
                        .name(publication.getAuthor().getName())
                        .email(publication.getAuthor().getEmail())
                        .build())
                .title(publication.getTitle())
                .description(publication.getDescription())
                .content(publication.getContent())
                .tags(publication.getTags())
                .status(publication.getStatus())
                .viewCount(publication.getViewCount())
                .likeCount(publication.getLikeCount())
                .commentCount(publication.getCommentCount())
                .featuredImage(publication.getFeaturedImage())
                .readingTime(publication.getReadingTime())
                .createdAt(publication.getCreatedAt())
                .updatedAt(publication.getUpdatedAt())
                .publishedAt(publication.getPublishedAt())
                .isLikedByCurrentUser(isLiked)
                .build();
    }
}
